addons/website/models/digest.py

```diff
@@ -0,0 +1,64 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import fields, models
+
+
+class Digest(models.Model):
+    _inherit = 'digest.digest'
+
+    kpi_website_visitor_cnt = fields.Boolean('Visitors')
+    kpi_website_visitor_cnt_value = fields.Integer(compute='_compute_kpi_website_visitors_cnt_value')
+    kpi_tracked_page_view_cnt = fields.Boolean('Tracked page views')  # Non uniq
+    kpi_tracked_page_view_cnt_value = fields.Integer(compute='_compute_kpi_tracked_page_view_cnt_value')
+
+    def _compute_visitor_cnt(self, ctx):
+        website_ids, start, end = ctx
+        if len(website_ids) == 0:
+            return 0
+        # The query takes into account that some website_visitor are the same visitor through parent_id
+        self._cr.execute("""
+            SELECT count(distinct COALESCE(v.parent_id, v.id))
+                FROM website_visitor AS v
+                JOIN website_track AS t ON t.visitor_id = v.id AND t.visit_datetime >= %(start)s AND t.visit_datetime < %(end)s
+                WHERE v.website_id IN %(website_ids)s""", {'website_ids': website_ids, 'start': start, 'end': end})
+        return self._cr.fetchone()[0]
+
+    def _compute_page_view_cnt(self, ctx):
+        website_ids, start, end = ctx
+        if len(website_ids) == 0:
+            return 0
+        self._cr.execute("""
+            SELECT count(*)
+                FROM website_visitor AS v
+                JOIN website_track AS t ON t.visitor_id = v.id AND t.visit_datetime >= %(start)s AND t.visit_datetime < %(end)s
+                WHERE v.website_id IN %(website_ids)s""", {'website_ids': website_ids, 'start': start, 'end': end})
+        return self._cr.fetchone()[0]
+
+    def _compute_by_websites_gen(self, compute_kpi_by_ctx):
+        """ Compute a value for each record using compute_kpi_by_ctx and yield the record and the computed value.
+        compute_kpi_by_ctx receives a kpi parameter tuple (website_ids, start, end) and return the computed kpi.
+        """
+        # Following dicts are for memoization of partial results during the loop (lru_cache not usable)
+        website_ids_by_company = dict()
+        visitor_cnt_by_ctx = dict()
+        for record in self:
+            start, end, company = record._get_kpi_compute_parameters()
+            website_ids = website_ids_by_company.get(company.id) is not None or \
+                          website_ids_by_company.setdefault(company.id, tuple(
```

Comment:
Seems a bit complicated :D not sure we need the ordering, the mapped on id can be replaced by ids, and wondering for the tuple (maybe afterwards if you have some sql you can tuple it when required)... becoming something like
```
website_ids_by_company.setdefault(
    company.id,
    self.env['website'].search([('company_id', '=', company.id)]).ids
)
``` ?

Reply:
I have improved the readability of that one a bit but the goal was to not execute the search if it was already computed (already in the website_ids_by_company). I haven't found a way to do memoization in the odoo code and in python library (defaultdict doesn't provide the key in the lambda that allows to create a missing value). I hope it is enough...